### PR TITLE
[ Xedra Evolved ] Vampvirus

### PR DESCRIPTION
**Pull request by Maleclypse**
#### Summary
Mods "Vampire Virus in Xedra Evolved"

#### Purpose of change
Vampires, vampires infect player, player succumbs to horrific disease and becomes the monster. Or player creates antidote and becomes their own hero.

#### Describe the solution
Adds a vampire virus, drugs to treat it, traits to tier it, monsters to inflict it.

#### Describe alternatives you've considered
None, literally none.

#### Testing
Everything works!

#### Additional context
No substitutions for human blood will be accepted.

### CODE REVIEW

**Comment by sonphantrung** [comment]
Hmm, will there be vampire mutagens in the future?

**Comment by Maleclypse** [comment]
> Hmm, will there be vampire mutagens in the future?

Probably not.

File: data/mods/Xedra_Evolved/effects/effects.json [hunk 1]
```diff
@@ -127,5 +127,88 @@
...
+    "blood_analysis_description": "Heightened levels of organosulfur compounds, mercury, and silver.",
+    "max_duration": 86400,
+    "dur_add_perc": 25
+  },
+  {
+    "type": "effect_type",
+    "id": "blood_quenched",
+    "rating": "good",
+    "blocks_effects": [ "vampire_virus" ],
+    "blood_analysis_description": "Heightened levels of organosulfur compounds, mercury, and silver.",
```
Comment (TheShadowFerret) [inline data/mods/Xedra_Evolved/effects/effects.json:145]: I don't think blood is made of these 🤔

**Comment by TheShadowFerret** [comment]
Will there be a scenario/profession/whatever that allows us to start infected with vampirism? 👀

**Comment by Maleclypse** [comment]
> Will there be a scenario/prof/whatever that allows us to start infected with vampirism? 👀

Probably... :)

**Comment by Maleclypse** [comment]
@dseguin There are two errors in the c++ that I'm not 100% certain what to do with.  I appreciate all your help so far and I think I'm pretty close to being able to test this in game.

**Comment by dseguin** [comment]
Try:
```diff
diff --git a/src/mattack_actors.cpp b/src/mattack_actors.cpp
index 02ee97353b..34c32c56b9 100644
--- a/src/mattack_actors.cpp
+++ b/src/mattack_actors.cpp
@@ -46,6 +46,7 @@ static const efftype_id effect_stunned( "stunned" );
 static const efftype_id effect_targeted( "targeted" );
 static const efftype_id effect_was_laserlocked( "was_laserlocked" );
 static const efftype_id effect_zombie_virus( "zombie_virus" );
+static const efftype_id effect_vampire_virus( "vampire_virus" );
 
 static const skill_id skill_throw( "throw" );
 
@@ -671,8 +672,7 @@ void bite_actor::on_damage( monster &z, Creature &target, dealt_damage_instance
         if( z.has_flag( MF_DEADLY_VIRUS ) && !target.has_effect( effect_zombie_virus ) ) {
             target.add_effect( effect_zombie_virus, 1_turns, bodypart_str_id::NULL_ID(), true );
         } else if( z.has_flag( MF_VAMP_VIRUS ) && !target.has_trait( trait_VAMPIRE ) ) {
-            dialogue d( get_talker_for( get_avatar() ), nullptr );
-            effect_on_condition_VAMP_VIRUS->activate( d );
+            target.add_effect( effect_vampire_virus, 1_turns, bodypart_str_id::NULL_ID(), true );
         }
     }
```
This will get it compiling, and it adds the effect directly instead of calling an EOC. Haven't tested it myself, but that will at least let you try it out

File: data/mods/Xedra_Evolved/effects/effects.json [hunk 2]
```diff
@@ -127,5 +127,83 @@
...
+    "dur_add_perc": 25
+  },
+  {
+    "type": "effect_type",
+    "id": "vampire_virus",
+    "name": [ "Wan", "Reddened Eyes", "Bloodthirst", "Terminal Porphyria", "Homo sapiens homovorus" ],
+    "desc": [
+      "Your complexion is wan and spent.",
+      "Your skin is pale and your eyes are thoroughly bloodshot.",
+      "Your skin is pale and have turned dark red.  You feel a constant hunger, like the food you eat is no longer nourishing you.",
```
Comment (TheShadowFerret) [inline data/mods/Xedra_Evolved/effects/effects.json:189]:
```suggestion
      "Your skin is pale and your eyes have turned dark red.  You feel a constant hunger, like the food you eat is no longer nourishing you.",
```
Pale dark red 🤔

**Comment by Maleclypse** [comment]
@Ramza13 Sorry to bother I have run into an error I'm not quite sure how to fix in EOCs.

**Comment by Maleclypse** [comment]
@dseguin I've mirrored your EOC for caffeine addiction and it's giving me an error that I don't understand.  If you have a second can you review the addiction eoc I've added here?

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+      ]
+    }
+  },
+  {
+    "type": "effect_on_condition",
+    "id": "EOC_vampire_virus_blood_needs",
+    "recurrence": [ "8 hours", "36 hours 15 minutes" ],
+    "condition": {
+      "and": [
+        { "compare_int": [ { "u_val": "vitamin", "name": "blood" }, "=<", { "const": 450 } ] },
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:209]:
```suggestion
        { "compare_int": [ { "u_val": "vitamin", "name": "blood" }, "<=", { "const": 450 } ] },
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+          "effect": {
+            "run_eocs": {
+              "id": "eoc_blood_hunger",
+              "condition": { "u_has_trait": "VAMPIRE" },
+              "effect": { "u_message": "You daydream of drinking blood from a chalice." }
+            }
+          }
+        },
+        {
+          "id": "EOC_VAMPVIRUS1",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:233]:
```suggestion
          "id": "EOC_VAMPVIRUS5",
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+          "condition": {
+            "and": [
+              { "u_has_effect": "vampire_virus", "intensity": 2 },
+              { "compare_int": [ { "u_val": "effect_intensity", "effect": "vampire_virus" }, "=", { "const": 2 } ] }
+            ]
+          },
+          "effect": {
+            "run_eocs": {
+              "id": "eoc_vampvirus_2_trait",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:242]:
```suggestion
              "id": "eoc_vampvirus_6_trait",
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+                { "u_add_trait": "VAMPIRE2" },
+                {
+                  "u_message": "With a popping sound several of your teeth fall out but when you put a finger into your mouth to feel the hole, there are new sharp fangs where you expected nothing."
+                }
+              ]
+            }
+          }
+        },
+        {
+          "id": "EOC_VAMPVIRUS2",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:255]:
```suggestion
          "id": "EOC_VAMPVIRUS6",
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+          "condition": {
+            "and": [
+              { "u_has_effect": "vampire_virus", "intensity": 3 },
+              { "compare_int": [ { "u_val": "effect_intensity", "effect": "vampire_virus" }, "=", { "const": 3 } ] }
+            ]
+          },
+          "effect": {
+            "run_eocs": {
+              "id": "eoc_vampvirus_3_trait",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:264]:
```suggestion
              "id": "eoc_vampvirus_7_trait",
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+              "condition": { "u_has_trait": "VAMPIRE3" },
+              "effect": { "u_message": "You feel a gnawing in your belly and a pressure in your mouth, as if your teeth can't all fit." },
+              "false_effect": [ { "u_add_trait": "VAMPIRE3" }, { "u_message": "Your skin seems to have become unnaturally pale." } ]
+            }
+          }
+        },
+        {
+          "id": "EOC_VAMPVIRUS3",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:272]:
```suggestion
          "id": "EOC_VAMPVIRUS7",
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+          "condition": {
+            "and": [
+              { "u_has_effect": "vampire_virus", "intensity": 4 },
+              { "compare_int": [ { "u_val": "effect_intensity", "effect": "vampire_virus" }, "=", { "const": 4 } ] }
+            ]
+          },
+          "effect": {
+            "run_eocs": {
+              "id": "eoc_vampvirus_4_trait",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:281]:
```suggestion
              "id": "eoc_vampvirus_8_trait",
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+              "condition": { "u_has_trait": "VAMPIRE4" },
+              "effect": { "u_message": "You feel a gnawing in your belly and a pressure in your mouth, as if your teeth can't all fit." },
+              "false_effect": [ { "u_add_trait": "VAMPIRE4" }, { "u_message": "Your skin seems to have become unnaturally pale." } ]
+            }
+          }
+        },
+        {
+          "id": "EOC_VAMPVIRUS4",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:289]:
```suggestion
          "id": "EOC_VAMPVIRUS8",
```

File: data/mods/Xedra_Evolved/effects/vampvirus.json [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+          "condition": {
+            "and": [
+              { "u_has_effect": "vampire_virus", "intensity": 5 },
+              { "compare_int": [ { "u_val": "effect_intensity", "effect": "vampire_virus" }, "=", { "const": 5 } ] }
+            ]
+          },
+          "effect": {
+            "run_eocs": {
+              "id": "eoc_vampvirus_5_trait",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/effects/vampvirus.json:298]:
```suggestion
              "id": "eoc_vampvirus_9_trait",
```

File: data/mods/Xedra_Evolved/monster_special_attacks/monster_special_attacks.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    "effect": "effect_on_condition",
+    "effect_str": "EOC_vampire_bite",
+    "damage_type": "stab",
+    "min_damage": 15,
+    "max_damage": 25,
+    "damage_increment": 1.0,
+    "max_level": 10,
+    "min_range": 1,
+    "max_range": 1,
+    "effect": "attack",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/monster_special_attacks/monster_special_attacks.json:18]:
```suggestion
```
Duplicated `"effect"` field

File: data/mods/Xedra_Evolved/mutations/mutations.json [hunk 13]
```diff
@@ -46,5 +46,17 @@
...
+    "type": "mutation",
+    "id": "VAMPIRE",
+    "name": { "str": "Infected with the vampire disease." },
+    "points": 0,
+    "player_display": false,
+    "description": "You have developed a thirst for the blood of your fellow humans.",
+    "starting_trait": false,
+    "purifiable": false,
+    "valid": false,
+    "learn_spell": "consume_human_blood"
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/mutations/mutations.json:60]:
```suggestion
    "spells_learned": [ [ "consume_human_blood", 1 ] ]
```

File: src/monstergenerator.cpp [hunk 14]
```diff
@@ -211,6 +211,7 @@ std::string enum_to_string<m_flag>( m_flag data )
         case MF_ATTACK_UPPER: return "ATTACK_UPPER";
         case MF_ATTACK_LOWER: return "ATTACK_LOWER";
         case MF_DEADLY_VIRUS: return "DEADLY_VIRUS";
+        case MF_VAMP_VIRUS: return "VAMPIRE_VIRUS";
```
Comment (dseguin) [inline src/monstergenerator.cpp:214]:
```suggestion
        case MF_VAMP_VIRUS: return "VAMP_VIRUS";
```

File: data/mods/Xedra_Evolved/addictions.json [hunk 15]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "type": "addiction_type",
+    "id": "human_blood",
+    "name": "Blood Hunger",
+    "type_name": "blood_hunger",
+    "description": "You may find it difficult to sleep without medication.",
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/addictions.json:7]:
```suggestion
    "description": "You suffer from a quite unnatural thirst for the blood of your fellow man.  Unfortunately, the supply has shrunk by about 99% from a few months ago.",
```

**Comment by Ramza13** [comment]
> @Ramza13 Sorry to bother I have run into an error I'm not quite sure how to fix in EOCs.

Sorry I was on vacation, what seems to be the problem?

**Comment by Maleclypse** [comment]
> > @Ramza13 Sorry to bother I have run into an error I'm not quite sure how to fix in EOCs.
> 
> Sorry I was on vacation, what seems to be the problem?

I did <= instead of =< if I remember from Dseguin's fix they showed me.  I had many other errors too, that were less syntax issues and more counting errors.

**Comment by Maleclypse** [comment]
Hey @Ramza13 I'm double checking against your PR and I think I've done these two roll remainder eoc's backwards?   I think you added functionality to basically have what I'm using as the secondary eoc condition so that there's a more elegant way to achieve what I'm trying to do? 

```
"effect": {
            "run_eocs": {
              "id": "eoc_vampvirus_1_trait",
              "condition": { "u_has_trait": "VAMPIRE" },
              "effect": {
                "run_eocs": [
                  {
                    "id": "EOC_VAMPVIRUS_TRAITS",
                    "condition": { "and": [ { "u_has_trait": "VAMPIRE" }, { "u_has_trait": "EYEGLEAM" }, { "u_has_trait": "STAMINAFORBLOOD" } ] },
                    "effect": "EOC_VAMPVIRUS1",
                    "false_effect": [
                      { "u_roll_remainder": [ "EYEGLEAM", "STAMINAFORBLOOD" ], "type": "mutation" },
                      { "u_message": "You realize a new ability as a result of the changes your body is undergoing." }
                    ]
                  }
```
Because this is a progressive disease I want the first check to see if you have a VAMPIRE trait and if you don't that VAMPIRE trait is added.  Then if you do have the VAMPIRE trait it adds from the list and if you have all the listed traits it checks for the next level of VAMPIRE.

**Comment by Maleclypse** [comment]
@dseguin Everything else is testing great I need to add messages in some places but the blood treatment eoc isn't reducing the virus effect?
```
"run_eocs": [
          {
            "id": "EOC_REDUCE_VAMPVIRUS",
            "condition": {
              "and": [
                { "u_has_effect": "vampire_virus", "intensity": 1 },
                { "compare_int": [ { "u_val": "effect_intensity", "effect": "vampire_virus" }, "=", { "const": 1 } ] }
              ]
            },
            "effect": [
              { "u_lose_effect": "vampire_virus" },
              { "u_message": "The light seems less bright and the air fresher.", "type": "good" }
            ]
          },
```

**Comment by Maleclypse** [comment]
Also the `"flags": [ "CANNIBAL" ]` isn't preventing feels bad messages though it is preventing morale damage

**Comment by dseguin** [comment]
> Everything else is testing great I need to add messages in some places but the blood treatment eoc isn't reducing the virus effect?

I'm not really sure how you're testing this or how you want it to work...

Just throwing out guesses:
- Do you mean that the EOC doesn't trigger? I don't see `EOC_CONSUMED_BLOOD_TREATMENT` referenced anywhere else and it's non-recurring so I don't think it will trigger by itself
- It looks like your two conditions are redundant, you could probably just get away with:
```
"condition": { "compare_int": [ { "u_val": "effect_intensity", "effect": "vampire_virus" }, "=", { "const": 1 } ] }
```
- Just tested `EOC_REDUCE_VAMPVIRUS` when the effect has 1 intensity, seems to work:
msg
- Currently when you consume the "blood treatment" item, it just gives you the effect without triggering the EOC. Not sure what the intended behaviour is but it looks like it should trigger `EOC_CONSUMED_BLOOD_TREATMENT` instead of giving the effect in the `use_action`

<hr>

> Also the "flags": [ "CANNIBAL" ] isn't preventing feels bad messages though it is preventing morale damage

Same deal as above. Is this with the `VAMPIRE` mutation? or one of the other mutations? Which messages do you mean (consume, dissect, etc)?

Just tested with the basic `VAMPIRE` mutation (which is the only one with the flag). Looks fine, no feels bad messages:

msg2

**Comment by Maleclypse** [comment]
> 

Oh just realized the cannibal flag wasn't attached to my first character because by the time the vamp virus escalation eoc fired I'd gone up to the second tier of vampire skipping the first.   I've corrected that in that each level of the disease progression should now last 33 days roughly.  I should probably also add the cannibal flag in each Vampire mutation just in case.  

I'm trying to decide if I should get rid of the consume blood spell and just have the character eat blood normally for the vitamins, but I'm concerned with how I'd accomplish that without breaking non-vampire characters by making them gain excess blood_vitamin they can't get rid of.

**Comment by Maleclypse** [comment]
I don't know what I've done but the escalation eoc doesn't seem to be running any more? ```  {
    "type": "effect_on_condition",
    "id": "EOC_VAMPIRES",
    "recurrence": [ { "global_val": "ps_min_woc", "default": "1 days" }, { "global_val": "ps_max_woc", "default": "2 days" } ],
    "global": true,
    "effect": { "run_eocs": "EOC_vampire_virus_escalation" }
  },
  {
    "type": "effect_on_condition",
    "id": "EOC_vampire_virus_escalation",
    "recurrence": [ "48 hours", "72 hours 15 minutes" ],```
I swear I had this running last night and then I looked at it this morning and was like "I think this is missing a global timer" and now it's not working

**Comment by Maleclypse** [comment]
Found my error :)

**Comment by DamienRoyan** [comment]
Just curious, by "succumbs to horrific disease and becomes the monster", does this mean that my character will die when it completes or will I continue to play, being turned into a monster "against" my will

**Pull request merged**